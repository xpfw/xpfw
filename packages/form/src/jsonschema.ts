
export declare interface ValidationKeywords {
  [prop: string]: boolean
}
export declare interface Args {
  ref: boolean
  aliasRef: boolean
  topRef: boolean
  titles: boolean
  defaultProps: boolean
  noExtraProps: boolean
  propOrder: boolean
  typeOfKeyword: boolean
  required: boolean
  strictNullChecks: boolean
  ignoreErrors: boolean
  out: string
  validationKeywords: string[]
  include: string[]
  excludePrivate: boolean
  uniqueNames: boolean
  rejectDateType: boolean
  id: string
}
export declare type PartialArgs = Partial<Args>
export declare type PrimitiveType = number | boolean | string | null
export declare interface JSONSchemaDefinition {
  $ref?: string
  $schema?: string
  $id?: string
  description?: string
  allOf?: JSONSchemaDefinition[]
  oneOf?: JSONSchemaDefinition[]
  anyOf?: JSONSchemaDefinition[]
  title?: string
  type?: string | string[]
  definitions?: {
      [key: string]: any;
  }
  format?: string
  items?: JSONSchemaDefinition | JSONSchemaDefinition[]
  minItems?: number
  additionalItems?: {
      anyOf: JSONSchemaDefinition[];
  } | JSONSchemaDefinition
  enum?: PrimitiveType[] | JSONSchemaDefinition[]
  default?: PrimitiveType | Object
  additionalProperties?: JSONSchemaDefinition | boolean
  required?: string[]
  propertyOrder?: string[]
  properties?: {
      [key: string]: JSONSchemaDefinition | undefined;
  }
  defaultProperties?: string[]
  patternProperties?: {
      [pattern: string]: JSONSchemaDefinition;
  }
  typeof?: "function"
}

export interface ISelectItem {
  /**
   * will be displayed in the user interface
   */
  label: string
  /**
   * will be used to set the value
   */
  value: any
}

export type SelectOptionGetter = (state: any) => ISelectItem[]

export type SelectOptionsField = ISelectItem[] | SelectOptionGetter

export interface IFieldVisibilityDef {
  create?: boolean
  update?: boolean
  find?: boolean
  remove?: boolean
}

export type ModifyFunction = (value: any, schema: ExtendedJSONSchema, method: string) => Promise<any>

export declare interface ExtendedJSONSchema extends JSONSchemaDefinition {
  /**
   * `selectOptions` allows rendering a `<select>` in react or a `Picker` in react-native.
   *  Set the `type`-key to `number` or `string` and set the `format`-key to `select` in the JSONSchemaDefintion.
   *  It's value should be either an object or a function returning an object.
   * The object is expected to have a `value`-key which will be used to set the value
   * as well as `label`-key which will be displayed in the user interface.
   */
  selectOptions?: SelectOptionsField
  /**
   * Apply a theme to a field
   * Could for example be used to render custom components per field
   */
  theme?: string
  collection?: string
  multiCollection?: string[]
  relationship?: {
    /**
     * Collection to search in if Field has type RelationshipSingle or RelationshipMany
     */
    collection?: string
    /**
     * Path to get name from if has type RelationshipSingle or RelationshipMany
     */
    namePath?: string
    /**
     * Path to get id from if has type RelationshipSingle or RelationshipMany
     */
    idPath?: string
    /**
     * if set to true it wil auto select an item in relationship search if it is the only one found
     */
    autoSelect?: boolean
    /**
     * if set to true the query will contain a $nin on `dataOptions.idPath`
     */
    filterOutSelected?: boolean
  }
  /**
   * Gives the ability to modify an object before submission via create, patch or find
   */
  modify?: ModifyFunction | ModifyFunction[]
  /**
   * Gives the ability to hide the field in situations such as CRUD-environments like useCreate
   */
  hide?: IFieldVisibilityDef
}
