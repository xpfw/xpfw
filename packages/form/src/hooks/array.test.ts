import { JSONSchemaDefinition } from "../jsonschema"
import FormStore from "../store/form"
import useArrayHelper from "./array"

describe("Form Hooks Test", () => {
  it("useArrayHelper", () => {
    const valueToSet = "myStringValue"
    let arrayHelper = useArrayHelper(valueToSet)
    expect(FormStore.formData).toMatchSnapshot("Before anything")
    arrayHelper.fields[0].increaseSize()
    expect(FormStore.formData).toMatchSnapshot("After calling increase size on subfield")
    arrayHelper.increaseSize()
    arrayHelper.increaseSize()
    expect(FormStore.formData).toMatchSnapshot("After calling increase size on root")
    arrayHelper = useArrayHelper(valueToSet)
    FormStore.setValue(arrayHelper.fields[1].mapTo, " set specific value ")
    expect(FormStore.formData).toMatchSnapshot("After Using subfield map to with set value")
    arrayHelper.fields[1].decreaseSize()
    expect(FormStore.formData).toMatchSnapshot("After Using subfield Decrease size")
    arrayHelper.decreaseSize()
    expect(FormStore.formData).toMatchSnapshot("After Using Global Decrease size")
  })
})
