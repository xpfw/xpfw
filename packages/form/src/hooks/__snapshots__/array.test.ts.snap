// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Hooks Test useArrayHelper: After Using Global Decrease size 1`] = `
Object {
  "myStringValue": Array [
    undefined,
    undefined,
    undefined,
  ],
}
`;

exports[`Form Hooks Test useArrayHelper: After Using subfield Decrease size 1`] = `
Object {
  "myStringValue": Array [
    undefined,
    undefined,
    undefined,
  ],
}
`;

exports[`Form Hooks Test useArrayHelper: After Using subfield map to with set value 1`] = `
Object {
  "myStringValue": Array [
    undefined,
    " set specific value ",
    undefined,
    undefined,
  ],
}
`;

exports[`Form Hooks Test useArrayHelper: After calling increase size on root 1`] = `
Object {
  "myStringValue": Array [
    undefined,
    undefined,
    undefined,
    undefined,
  ],
}
`;

exports[`Form Hooks Test useArrayHelper: After calling increase size on subfield 1`] = `
Object {
  "myStringValue": Array [
    undefined,
    undefined,
  ],
}
`;

exports[`Form Hooks Test useArrayHelper: Before anything 1`] = `Object {}`;
